added forgot-password

YassenAhmed-gif opened the pull request.

YassenAhmed-gif committed:
added forgot-password

YassenAhmed-gif committed:
feat: Add reset password functionality

This commit introduces the functionality to reset a user's password. It includes:

- API client methods for verifying the reset code and changing the password.
- Data sources, repositories, and use cases for handling reset code verification and password change.
- ViewModels for the reset code and change password screens.
- DTOs and entities for reset code and password change responses.
- Necessary DI configurations for the new components.
- Placeholder screens for entering the reset code and changing the password.

YassenAhmed-gif committed:
feat: Add reset password functionality

This commit introduces the functionality to reset a user's password. It includes:

- API client methods for verifying the reset code and changing the password.
- Data sources, repositories, and use cases for handling reset code verification and password change.
- ViewModels for the reset code and change password screens.
- DTOs and entities for reset code and password change responses.
- Necessary DI configurations for the new components.
- Placeholder screens for entering the reset code and changing the password.

john-safwat commented on lib/project_layers/Domain_layer/model/Forgot.dart:
```diff
@@ -0,0 +1,12 @@
+class Forgot {
```
miss leading class name please make class name more clear

john-safwat commented on lib/project_layers/Domain_layer/model/Reset.dart:
```diff
@@ -0,0 +1,11 @@
+class Reset {
```
also here is a miss leading class name please change it

john-safwat commented on lib/project_layers/Domain_layer/model/resetPassword.dart:
```diff
@@ -0,0 +1,11 @@
+class ResetPassword{
```
also here

john-safwat commented on lib/project_layers/Domain_layer/repositories/ForgotPasswordRepo.dart:
```diff
@@ -0,0 +1,6 @@
+
+import '../../api_layer/entities/ForgotPasswordEntity.dart';
+
+abstract class ForgotPasswordRepo{
```
use interface modifier in Repo to prevent any one from extending it

john-safwat commented on lib/project_layers/Domain_layer/repositories/ForgotPasswordRepo.dart:
```diff
@@ -0,0 +1,6 @@
+
+import '../../api_layer/entities/ForgotPasswordEntity.dart';
+
+abstract class ForgotPasswordRepo{
+  Future<ForgotPasswordEntity> sendEmail(String email);
+}
```
don't create class for each request merge all of this repos in one repo for auth feature

john-safwat commented on lib/project_layers/Domain_layer/usecase/ChangePasswordUseCase.dart:
```diff
@@ -0,0 +1,12 @@
+
+  ChangePasswordUseCase(this._repo);
+
+  Future<ResetPasswordEntity> change(String email, String password){
```
make function name clear

john-safwat commented on lib/project_layers/api_layer/api_client/api_module.dart:
```diff
@@ -0,0 +1,33 @@
+        receiveTimeout: const Duration(seconds: 20),
+      )
+    );
+    dio.interceptors.add(PrettyDioLogger(
+        requestHeader: true,
+        requestBody: true,
+        responseBody: true,
+        responseHeader: false,
+        error: true,
+        compact: true,
+        maxWidth: 90,
+        enabled: kDebugMode,
+    )
```
split the creation of pretty dio logger to another function

john-safwat commented on lib/project_layers/api_layer/dataSource/ForgotPasswordRemoteDataSourceimpl.dart:
```diff
@@ -0,0 +1,29 @@
+    try {
+      var response = await _client.postEmail({'email': email});
+      return response.toForgot();
+    } on DioException catch (e) {
+      // It's good practice to log the error or handle specific DioError types
+      print("DioException in sendEmail: ${e.message}");
+      print("Response data: ${e.response?.data}");
+      throw Exception("Failed to send email: ${e.message}");
+    } catch(e) {
+      print("Unknown error in sendEmail: $e");
+      throw Exception("An unknown error occurred: $e");
+    }
+  }
```
don't user prints in you code use debugger or loggers instead

john-safwat reviewed (changes requested):
- please modify you code to make all this repos in one single repo for auth feature not every single function need to be placed in repo 
- also make single data source instead of all this datasource all have same responsibility of user authentication 
- please make sure the class names , function name , variables all to be clear and readable 
- avoid using hard coded strings in your code
- avoid prints in your code

YassenAhmed-gif commented:
Re: Requested Changes for Auth Feature Refactoring

Hi John,

Thank you for your feedback on the pull request. I’ve reviewed your
requested changes regarding consolidating the repositories and data sources
for the auth feature, as well as improving code clarity. Here’s my
understanding of the required modifications:

1. **Consolidate Repositories & Data Sources**
   - I’ll merge all auth-related repositories into a single
`AuthRepository` to centralize authentication logic.
   - Similarly, I’ll unify the data sources into a single `AuthDataSource`
to avoid redundancy in user authentication operations.

2. **Improve Code Readability**
   - Class, function, and variable names will be updated to follow clear,
descriptive conventions (e.g., `AuthRepository` instead of `UserRepo`).
   - Hardcoded strings will be replaced with constants or configuration
values.

3. **Remove Debugging Artifacts**
   - All `print` statements will be replaced with proper logging where
necessary.

4. **Single Responsibility Principle**
   - The `AuthRepository` will handle all auth-related functions (login,
logout, token refresh, etc.), while the `AuthDataSource` will manage data
operations (e.g., API/database calls).

I’ll start refactoring and share an updated PR once these changes are
implemented. Let me know if you’d like to discuss further.

Best regards,

في الأربعاء، 16 يوليو 2025 12:45 ص John Safwat ***@***.***>
كتب:

> ***@***.**** requested changes on this pull request.
>
>    - please modify you code to make all this repos in one single repo for
>    auth feature not every single function need to be placed in repo
>    - also make single data source instead of all this datasource all have
>    same responsibility of user authentication
>    - please make sure the class names , function name , variables all to
>    be clear and readable
>    - avoid using hard coded strings in your code
>    - avoid prints in your code
>
> ------------------------------
>
> In lib/project_layers/Domain_layer/model/Forgot.dart
> <https://github.com/Aswani20/online_exam_app/pull/4#discussion_r2208751848>
> :
>
> > @@ -0,0 +1,12 @@
> +class Forgot {
>
> miss leading class name please make class name more clear
> ------------------------------
>
> In lib/project_layers/Domain_layer/model/Reset.dart
> <https://github.com/Aswani20/online_exam_app/pull/4#discussion_r2208752570>
> :
>
> > @@ -0,0 +1,11 @@
> +class Reset {
>
> also here is a miss leading class name please change it
> ------------------------------
>
> In lib/project_layers/Domain_layer/model/resetPassword.dart
> <https://github.com/Aswani20/online_exam_app/pull/4#discussion_r2208753030>
> :
>
> > @@ -0,0 +1,11 @@
> +class ResetPassword{
>
> also here
> ------------------------------
>
> In lib/project_layers/Domain_layer/repositories/ForgotPasswordRepo.dart
> <https://github.com/Aswani20/online_exam_app/pull/4#discussion_r2208753940>
> :
>
> > @@ -0,0 +1,6 @@
> +
> +import '../../api_layer/entities/ForgotPasswordEntity.dart';
> +
> +abstract class ForgotPasswordRepo{
>
> use interface modifier in Repo to prevent any one from extending it
> ------------------------------
>
> In lib/project_layers/Domain_layer/repositories/ForgotPasswordRepo.dart
> <https://github.com/Aswani20/online_exam_app/pull/4#discussion_r2208755511>
> :
>
> > +
> +import '../../api_layer/entities/ForgotPasswordEntity.dart';
> +
> +abstract class ForgotPasswordRepo{
> +  Future<ForgotPasswordEntity> sendEmail(String email);
> +}
>
> don't create class for each request merge all of this repos in one repo
> for auth feature
> ------------------------------
>
> In lib/project_layers/Domain_layer/usecase/ChangePasswordUseCase.dart
> <https://github.com/Aswani20/online_exam_app/pull/4#discussion_r2208757440>
> :
>
> > @@ -0,0 +1,12 @@
> +import 'package:online_exam_app/project_layers/Domain_layer/repositories/ReretPasswordRepo.dart';
> +import 'package:online_exam_app/project_layers/api_layer/entities/ResetPasswordEntity.dart';
> +
> +class ChangePasswordUseCase{
> +  final ResetPasswordRepo _repo;
> +
> +  ChangePasswordUseCase(this._repo);
> +
> +  Future<ResetPasswordEntity> change(String email, String password){
>
> make function name clear
> ------------------------------
>
> In lib/project_layers/api_layer/api_client/api_module.dart
> <https://github.com/Aswani20/online_exam_app/pull/4#discussion_r2208761382>
> :
>
> > +    dio.interceptors.add(PrettyDioLogger(
> +        requestHeader: true,
> +        requestBody: true,
> +        responseBody: true,
> +        responseHeader: false,
> +        error: true,
> +        compact: true,
> +        maxWidth: 90,
> +        enabled: kDebugMode,
> +    )
>
> split the creation of pretty dio logger to another function
> ------------------------------
>
> In
> lib/project_layers/api_layer/dataSource/ForgotPasswordRemoteDataSourceimpl.dart
> <https://github.com/Aswani20/online_exam_app/pull/4#discussion_r2208763080>
> :
>
> > +    } on DioException catch (e) {
> +      // It's good practice to log the error or handle specific DioError types
> +      print("DioException in sendEmail: ${e.message}");
> +      print("Response data: ${e.response?.data}");
> +      throw Exception("Failed to send email: ${e.message}");
> +    } catch(e) {
> +      print("Unknown error in sendEmail: $e");
> +      throw Exception("An unknown error occurred: $e");
> +    }
> +  }
>
> don't user prints in you code use debugger or loggers instead
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Aswani20/online_exam_app/pull/4#pullrequestreview-3022418896>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BLKV7IRHVXEIOWBIHMTWASD3IVY7VAVCNFSM6AAAAACBNROPICVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZTAMRSGQYTQOBZGY>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

YassenAhmed-gif committed:
Refactor: Consolidate auth-related data sources and use cases

This commit refactors the authentication-related data sources and use cases into a more unified structure.

Key changes:
- Introduced `AuthRepo`, `AuthRemoteDataSource`, and `AuthUseCase` to centralize authentication logic.
- Removed individual data sources and use cases for forgot password, reset code, and change password.
- Renamed model classes for clarity (e.g., `Forgot` to `ForgotPasswordModel`).
- Updated DTOs to use the new model names.
- Injected `PrettyDioLogger` into `ApiModule` for better debugging.
- Updated view models and screens to use the new unified `AuthViewModel` and `AuthUseCase`.
